automation/check-doc.js

```diff
@@ -0,0 +1,90 @@
+ */
+
+const { stripIndent } = require('common-tags');
+const _ = require('lodash');
+const { fs } = require('mz');
+const path = require('path');
+const simplegit = require('simple-git/promise');
+
+/**
+ * Compare the timestamp of cli.markdown with the timestamp of staged files,
+ * issuing an error if cli.markdown is older. Besides the purpose of ensuring
+ * that cli.markdown is updated, it effectively also ensures that coffeelint
+ * is executed (via `npm run build` or `npm test`) on the developers laptop,
+ * so that there is at least a chance that the developer will spot any linter
+ * warnings (that could reveal bugs) sooner than later.  (The CI does not
+ * currently fail in case of coffeelint warnings.)
+ * If cli.markdown does not require updating and the developer cannot run
+ * `npm run build` on their laptop, the error message suggests a workaround
+ * using `touch`.
+ */
+async function checkBuildTimestamps() {
+	const git = simplegit();
+	const docFile = path.join('doc', 'cli.markdown');
+	const [docStat, gitStatus] = await Promise.all([
+		fs.stat(docFile),
+		git.status(),
+	]);
+	const stagedFiles = _.uniq([
+		...gitStatus.created,
+		...gitStatus.staged,
+		...gitStatus.renamed.map(o => o.to),
+	])
+		// select only staged files that start with lib/ or typings/
+		.filter(f => f.match(/^(lib|typings)[/\\]/));
+
+	const fStats = await Promise.all(stagedFiles.map(f => fs.stat(f)));
+	fStats.forEach((fStat, index) => {
+		if (fStat.mtimeMs > docStat.mtimeMs) {
+			const fPath = stagedFiles[index];
+			throw new Error(stripIndent`
+				--------------------------------------------------------------------------------
+				ERROR: at least one staged file: "${fPath}"
+				has a more recent modification timestamp than the documentation file:
+				"${docFile}"
+
+				This probably means that \`npm run build\` or \`npm test\` have not been executed,
+				and this error can be fixed by doing so. Running \`npm run build\` or \`npm test\`
+				before commiting is currently a requirement (documented in the CONTRIBUTING.md
+				file) for three reasons:
+				1. Because it updates the CLI markdown documentation (in case any command-line
+				   options were updated, added or removed).
+				2. Because it reveals coffeelint warnings that the CI would currently ignore
+				   (in case any Coffeescript files were modified).
+				3. Given that balena-cli CI builds/tests are currently fairly slow, running
+				   tsc or resin-lint on the laptop catches Typescript type check errors much
+				   sooner and typically results in reduced overall waiting time.
+
+				If you need/wish to bypass this check without running \`npm run build\`, run
+				\`npx touch -am "${docFile}"\` and try again.
+				--------------------------------------------------------------------------------
+			`);
+		}
+	});
+}
+
+async function run() {
+	try {
+		await checkBuildTimestamps();
+	} catch (err) {
+		console.error(err.message);
+		process.exit(1);
```

Comment:
```suggestion
		process.exitCode = 1;
```
Whilst it doesn't usually happen, `libuv` can still be holding onto file handles after it's accessed files, and ensuring that it cleans itself up before exiting at the next event loop just makes things a bit better. `process.exitCode` ensures that the process still exists non-zero, but allows it to run the rest of the codepath in the current event loop and clean up first (and looking at the calling convention here, this is fine).